### PR TITLE
Added some fixes for minor CNV issues.

### LINKED ISSUE #3566: CNV OncotateSegments does not expose bootDiskInGb runtime parameter
This is often needed since the docker image is big.

Expose this parameter with a 20GB value

### LINKED ISSUE #4734: Add checks that number of matrix elements do not exceed max int in CreateReadCountPanelOfNormals.
Ran into this when trying to create a PoN with 100 samples x 100 bp bins = 1.27 * max int elements.

This currently causes issues when truncating outliers, at which point all elements are loaded into an array so that Percentiles can be naively computed, resulting in a `java.lang.NegativeArraySizeException`.

Solutions include: 1) simply throwing a message and failing early if the counts matrix is too large (perhaps recommend scattering by contig, see #4728), 2) changing the outlier truncation procedure to be more robust.

I'm not sure how important outlier truncation is to the SVD, as it remains to be evaluated, but for now we should be able to get around this with no code changes by simply disabling it (i.e., setting the relevant truncation percentile to 0).

Note that file I/O takes about an hour for this case.  Also note that this is probably on the extreme end of what we should expect to support on a single machine with all counts in memory, as the SVD is probably sufficiently good with 100 samples and 100 bp is on the order of the read length.  #4728 will get around this and also make downstream tasks complete faster in parallel, at the very small expense of reducing a few global parameters to per-contig parameters in the modeling step.

### LINKED ISSUE #4825: Ensure consistent output directory requirements in CNV tools.
The exception is:
```
Traceback (most recent call last):
  File "/home/shlee/anaconda3/envs/gatk/lib/python3.6/site-packages/gcnvkernel/io/io_commons.py", line 98, in assert_output_path_writable
    filehandle = open(filename, 'w')
PermissionError: [Errno 13] Permission denied: '/home/shlee/gcc/hc24_soohee1k_chr1-model/write_tester'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/shlee/cohort_denoising_calling.7832183760446168530.py", line 151, in <module>
    args.output_model_path)()
  File "/home/shlee/anaconda3/envs/gatk/lib/python3.6/site-packages/gcnvkernel/io/io_denoising_calling.py", line 28, in __init__
    io_commons.assert_output_path_writable(output_path)
  File "/home/shlee/anaconda3/envs/gatk/lib/python3.6/site-packages/gcnvkernel/io/io_commons.py", line 102, in assert_output_path_writable
    raise IOError("The output path \"{0}\" is not writeable".format(output_path))
OSError: The output path "/home/shlee/gcc/hc24_soohee1k_chr1-model" is not writeable
16:26:00.659 DEBUG ScriptExecutor - Result: 1
16:26:00.662 INFO  GermlineCNVCaller - Shutting down engine
[May 27, 2018 4:26:00 PM UTC] org.broadinstitute.hellbender.tools.copynumber.GermlineCNVCaller done. Elapsed time: 2,255.34 minutes.
Runtime.totalMemory()=8207728640
org.broadinstitute.hellbender.utils.python.PythonScriptExecutorException: 
python exited with 1
Command Line: python /tmp/shlee/cohort_denoising_calling.7832183760446168530.py --ploidy_calls_path=/home/shlee/gcnv/coverage_1k/hc24_soohee1kall_ploidy-calls --output_calls_path=/home/shlee/gcc/hc24_soohee1k_chr1-calls --modeling_interval_list=/tmp/shlee/intervals1147946183347323472.tsv --output_model_path=/home/shlee/gcc/hc24_soohee1k_chr1-model --enable_explicit_gc_bias_modeling=False --read_count_tsv_files /tmp/shlee/sample-04516283083315244626.tsv /tmp/shlee/sample-17497576995757363646.tsv /tmp/shlee/sample-21271002324475135098.tsv /tmp/shlee/sample-36985602309924438312.tsv /tmp/shlee/sample-44773997237633003175.tsv /tmp/shlee/sample-55563425618633690228.tsv /tmp/shlee/sample-66588087553393228850.tsv /tmp/shlee/sample-77682253251273859622.tsv /tmp/shlee/sample-85102983131590855033.tsv /tmp/shlee/sample-91827141265495425349.tsv /tmp/shlee/sample-103880243493985861748.tsv /tmp/shlee/sample-116680765766192550894.tsv /tmp/shlee/sample-12124239099888077446.tsv /tmp/shlee/sample-137882283449137233563.tsv /tmp/shlee/sample-143394886316760954899.tsv /tmp/shlee/sample-152734003651556671536.tsv /tmp/shlee/sample-164992531160419965633.tsv /tmp/shlee/sample-179066010714184030831.tsv /tmp/shlee/sample-186093298660325983992.tsv /tmp/shlee/sample-195129395964281545047.tsv /tmp/shlee/sample-205662400342674148944.tsv /tmp/shlee/sample-214794162996551180382.tsv /tmp/shlee/sample-221249794311427321965.tsv /tmp/shlee/sample-233870240722976827789.tsv --psi_s_scale=1.000000e-04 --mapping_error_rate=1.000000e-02 --depth_correction_tau=1.000000e+04 --q_c_expectation_mode=hybrid --max_bias_factors=5 --psi_t_scale=1.000000e-03 --log_mean_bias_std=1.000000e-01 --init_ard_rel_unexplained_variance=1.000000e-01 --num_gc_bins=20 --gc_curve_sd=1.000000e+00 --active_class_padding_hybrid_mode=50000 --enable_bias_factors=True --disable_bias_factors_in_active_class=False --p_alt=1.000000e-06 --cnv_coherence_length=1.000000e+04 --max_copy_number=5 --p_active=0.010000 --class_coherence_length=10000.000000 --learning_rate=5.000000e-02 --adamax_beta1=9.000000e-01 --adamax_beta2=9.900000e-01 --log_emission_samples_per_round=50 --log_emission_sampling_rounds=10 --log_emission_sampling_median_rel_error=5.000000e-03 --max_advi_iter_first_epoch=100 --max_advi_iter_subsequent_epochs=100 --min_training_epochs=10 --max_training_epochs=50 --initial_temperature=2.000000e+00 --num_thermal_epochs=20 --convergence_snr_averaging_window=500 --convergence_snr_trigger_threshold=1.000000e-01 --convergence_snr_countdown_window=10 --max_calling_iters=10 --caller_update_convergence_threshold=1.000000e-03 --caller_admixing_rate=7.500000e-01 --disable_caller=false --disable_sampler=false --disable_annealing=false
        at org.broadinstitute.hellbender.utils.python.PythonExecutorBase.getScriptException(PythonExecutorBase.java:75)
        at org.broadinstitute.hellbender.utils.runtime.ScriptExecutor.executeCuratedArgs(ScriptExecutor.java:126)
        at org.broadinstitute.hellbender.utils.python.PythonScriptExecutor.executeArgs(PythonScriptExecutor.java:170)
        at org.broadinstitute.hellbender.utils.python.PythonScriptExecutor.executeScript(PythonScriptExecutor.java:151)
        at org.broadinstitute.hellbender.utils.python.PythonScriptExecutor.executeScript(PythonScriptExecutor.java:121)
        at org.broadinstitute.hellbender.tools.copynumber.GermlineCNVCaller.executeGermlineCNVCallerPythonScript(GermlineCNVCaller.java:450)
        at org.broadinstitute.hellbender.tools.copynumber.GermlineCNVCaller.doWork(GermlineCNVCaller.java:288)
        at org.broadinstitute.hellbender.cmdline.CommandLineProgram.runTool(CommandLineProgram.java:134)
        at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMainPostParseArgs(CommandLineProgram.java:179)
        at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:198)
        at org.broadinstitute.hellbender.Main.runCommandLineProgram(Main.java:160)
        at org.broadinstitute.hellbender.Main.mainEntry(Main.java:203)
        at org.broadinstitute.hellbender.Main.main(Main.java:289)
```
Posting this as per @samuelklee's request.

### LINKED ISSUE #5382: gCNV WDLs should clean up intermediate files and directories.
Due to the way the calling process is sharded upstream ... the current wdl expends 80-90% of the time copying files over ... so a job that takes around 1h only 10 minutes are expend in running the GATK Tool, the result is used to stage the input files. 

For example if the genome intervals where shared into 600 ~ different batches this results in 3600 ~ files transferred one by one using their own gsutil command. The reason why these are not batched is in part because files share name and multi-file gsutil cp does not provide the means to indicate indepedendent destination names for each input file. Recursive copy of a parent directory would drag the information from all the samples when the each task just deals with one.

### LINKED ISSUE #5591: Output MAD values (not just text files) in the Somatic CNV WDL
### Tool(s) or class(es) involved
`PlotDenoisedCopyRatios`
`cnv_somatic_pair_workflow.wdl`

### Description
Currently, the output of the pair workflow only outputs files for the MAD values.  (e.g. `File standardized_MAD_tumor = PlotDenoisedCopyRatiosTumor.standardized_MAD`)
We should add outputs that are the actual data type (e.g. `Float standardized_MAD_tumor = read_float(PlotDenoisedCopyRatiosTumor.standardized_MAD)`)

Do this for all MAD values in the tumor and normal, not just the example above

### LINKED ISSUE #5744: Harden ScatterIntervals task in CNV WDL.
@LeeTL1220 mentioned having difficulty running gCNV on REBC WGS normals.  As far as I can tell, other than runs failing due to quota or FC issues, there were some runs that failed because the WDL version and the Docker version were not in sync.

Specifically, these runs failed because the WDL was more recent than the Docker, causing the ScatterIntervals task (which was running an out-of-date IntervalListTools in the latter) to "fail silently".  The behavior upon failure, for reasons due to the somewhat awkward format of the IntervalListTools output, is to output just a single shard by simply copying the original intervals list; this is handled in bash.  This single-shard run then failed due to OOM in the gCNV step.

Just to be clear, everything typically works fine when the versions are in sync.  But you could imagine that even then IntervalListTools could fail for other reasons, in which case we'd probably fail misleadingly at the gCNV step again.  So let's modify the WDL so we fail at the appropriate place.

Apologies to @asmirnov239, who I think pointed this weirdness out in the original PR review.

### PR DESCRIPTION
**Pull request by samuelklee**
-Cleaned up intermediate files in gCNV WDL.  Closes #5382.
-Added output of MAD values as floats in somatic CNV WDL.  Closes #5591.
-Exposed boot disk space for Oncotator in somatic CNV WDL.  Closes #3566.
-Added check to skip outlier truncation if number of matrix elements exceeds Integer.MAX_VALUE in CreateReadCountPanelOfNormals.  Closes #4734.
-Fixed some issues concerning intervals in DetermineGermlineContigPloidy documentation.
-Miscellaneous boy scout activities.
-Fixed non-kebab-case argument in CollectAllelicCountsSpark and other minor issues.  See #5478.
-Improved consistency of style and input/output validation across CNV tools.  Closes #4825.  Closes #5744.

### CODE REVIEW

**Comment by samuelklee** [comment]
Also filed #5749.

File: scripts/cnv_wdl/cnv_common_tasks.wdl [hunk 1]
```diff
@@ -301,28 +301,39 @@ task ScatterIntervals {
...
     command <<<
         set -e
+        # IntervalListTools will fail if the output directory does not exist, so we create it
         mkdir ${output_dir_}
         export GATK_LOCAL_JAR=${default="/root/gatk.jar" gatk4_jar_override}
-        
-        {
-            >&2 echo "Attempting to run IntervalListTools..."
+
+        # IntervalListTools behaves differently when scattering to a ssingle or multiple shards, so we do some handling in bash
```
Comment (samuelklee) [inline scripts/cnv_wdl/cnv_common_tasks.wdl]: ssingle -> single

**Comment by samuelklee** [comment]
@mwalker174 can you take a look?  There are some changes to how output directories are treated in the last commit that should perhaps be documented in the release notes, otherwise most changes are very minor.  Also tried to do a bit of polishing for style, but we can probably go back and do more.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/arguments/CopyNumberArgumentValidationUtils.java [hunk 2]
```diff
@@ -184,4 +187,82 @@ public static AnnotatedIntervalCollection validateAnnotatedIntervalsSubset(final
...
+    }
+
+    /**
+     * Validate that output files are writeable, whether or not they already exist.
+     */
+    public static void validateOutputFiles(final File ... outputs) {
+        Utils.nonNull(outputs);
+        for (final File output : outputs) {
+            Utils.nonNull(output);
+            if ((output.exists() && !output.canWrite()) || (!output.exists() && !output.getAbsoluteFile().getParentFile().canWrite())) {
```
Comment (mwalker174) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/arguments/CopyNumberArgumentValidationUtils.java]: Assuming this works on directories, if `output` is root (`/`) then `getParentFile()` returns null and this'll throw a null pointer exception.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/arguments/CopyNumberArgumentValidationUtils.java [hunk 3]
```diff
@@ -184,4 +187,82 @@ public static AnnotatedIntervalCollection validateAnnotatedIntervalsSubset(final
...
+                throw new UserException.CouldNotCreateOutputFile(output, ": The output file is not writeable.");
+            }
+        }
+    }
+
+    /**
+     * Validate that output directories are writeable.  If a directory does not exist, create it.
+     */
+    public static void validateAndPrepareOutputDirectories(final File ... outputs) {
```
Comment (mwalker174) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/arguments/CopyNumberArgumentValidationUtils.java]: I think you should combine this with `validateOutputFiles()` since `File` can be either a file or directory. Otherwise it seems that one could pass in dirs/files to the wrong function and get unexpected results.

File: scripts/cnv_wdl/cnv_common_tasks.wdl [hunk 4]
```diff
@@ -405,21 +422,22 @@ task PostprocessGermlineCNVCalls {
...
+            --contig-ploidy-calls contig-ploidy-calls \
             --sample-index ${sample_index} \
             --output-genotyped-intervals ${genotyped_intervals_vcf_filename} \
             --output-genotyped-segments ${genotyped_segments_vcf_filename}
 
-        rm -r CALLS_*
-        rm -r MODEL_*
+        rm -rf CALLS_*
+        rm -rf MODEL_*
+        rm -rf contig-ploidy-calls
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply scripts/cnv_wdl/cnv_common_tasks.wdl]: Just for those that might be running the WDLs locally.  I think @sooheelee might've brought this to my attention when she was generating tutorial materials at the command line.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CreateReadCountPanelOfNormals.java [hunk 5]
```diff
@@ -296,18 +293,22 @@ protected void runPipeline(final JavaSparkContext ctx) {
                 extremeSampleMedianPercentile, doImputeZeros, extremeOutlierTruncationPercentile, numEigensamplesRequested,
                 maximumChunkSize, ctx);
 
-        logger.info("Panel of normals successfully created.");
+        logger.info("CreateReadCountPanelOfNormals complete.");
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CreateReadCountPanelOfNormals.java]: Good call, didn't know that existed!  Done.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java [hunk 6]
```diff
@@ -81,9 +83,9 @@
  *      <p>The model subdirectory contains a snapshot of the inferred parameters of the coverage model, which may be
  *      used later for CNV calling in one or more similarly-sequenced samples as mentioned earlier. Optionally, the path
  *      to a previously obtained coverage model parametrization can be provided via the {@code model} argument
- *      in the COHORT mode, in which case, the provided parameters will be only used for model initialization and
+ *      in COHORT mode, in which case, the provided parameters will be only used for model initialization and
  *      a new parametrization will be generated based on the input count files. Furthermore, the genomic intervals are
- *      set to those used in creating the previous parameterization and interval-related arguments will be ignored.
+ *      set to those used in creating the previous parametrization and interval-related arguments will be ignored.
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java]: Just making a consistent choice across all CNV code.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CollectAllelicCountsSpark.java [hunk 7]
```diff
@@ -67,15 +60,38 @@
...
+
     @Override
     protected void processAlignments(JavaRDD<LocusWalkerContext> rdd, JavaSparkContext ctx) {
+        validateArguments();
+
         final SampleLocatableMetadata metadata = MetadataUtils.fromHeader(getHeaderForReads(), Metadata.Type.SAMPLE_LOCATABLE);
+        final SAMSequenceDictionary sequenceDictionary = getBestAvailableSequenceDictionary();
+        //this check is currently redundant, since the master dictionary is taken from the reads;
+        //however, if any other dictionary is added in the future, such a check should be performed
+        if (!CopyNumberArgumentValidationUtils.isSameDictionary(metadata.getSequenceDictionary(), sequenceDictionary)) {
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CollectAllelicCountsSpark.java]: Looking back at some pre-release chatter on Slack, I think we decided to punt on how strict CNV tools should be on dictionary validation.  I think `SAMSequenceDictionary.isSameDictionary` was typically too strict even for TCGA data.  This code was cribbed from that method, but with the `isSameSequence` comparison relaxed.  I guess what we are currently doing is equivalent to `IDENTICAL` in `SequenceDictionaryUtils`, so I'll go ahead and wrap that unless you want to change the validation behavior.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/arguments/CopyNumberArgumentValidationUtils.java (hunk 3 above)
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/arguments/CopyNumberArgumentValidationUtils.java]: Yeah, overall I'm not very happy with any of these validate input/output methods.  I originally had a single `validateOutputs` method, but the logic got a little ugly.  Let me try again.

In any case, we should probably overhaul all of the CNV Files over to Paths and revisit this issue.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/arguments/CopyNumberArgumentValidationUtils.java (hunk 2 above)
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/arguments/CopyNumberArgumentValidationUtils.java]: See below.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/arguments/CopyNumberArgumentValidationUtils.java (hunk 3 above)
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/arguments/CopyNumberArgumentValidationUtils.java]: Actually, I think the reason I went with two methods is that we can't check whether a non-existing File is a file or a directory, and we want to perform different actions in either case (check for write permission of the parent directory of the file and create the directory, respectively).  So we have to rely on calling code to correctly make the choice for us.  I think that the method names should suffice to direct a developer to the correct method, but I can change the argument names to make it even more explicit.

Might be a way to do this better, but I'm OK to punt until the overhaul.  There are other behaviors that we could decide upon (such as whether or not we want to create intermediate directories, which this code doesn't do).  Baby steps.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/DetermineGermlineContigPloidy.java [hunk 8]
```diff
@@ -380,25 +390,23 @@ private void writeSamplesByCoveragePerContig(final File samplesByCoveragePerCont
...
-        final String outputDirArg = Utils.nonEmpty(outputDir).endsWith(File.separator)
-                ? outputDir
-                : outputDir + File.separator;    //add trailing slash if necessary
+        final String outputDirArg = CopyNumberArgumentValidationUtils.addTrailingSlashIfNecessary(outputDir.getAbsolutePath());
         //note that the samples x coverage-by-contig table is referred to as "metadata" by gcnvkernel
         final List<String> arguments = new ArrayList<>(Arrays.asList(
-                "--sample_coverage_metadata=" + CopyNumberUtils.getCanonicalPath(samplesByCoveragePerContigFile),
-                "--output_calls_path=" + CopyNumberUtils.getCanonicalPath(outputDirArg + outputPrefix + CALLS_PATH_SUFFIX)));
+                "--sample_coverage_metadata=" + CopyNumberArgumentValidationUtils.getCanonicalPath(samplesByCoveragePerContigFile),
+                "--output_calls_path=" + CopyNumberArgumentValidationUtils.getCanonicalPath(outputDirArg + outputPrefix + CALLS_PATH_SUFFIX)));
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/DetermineGermlineContigPloidy.java]: python has a problem with non-canonical paths, which I believe `Paths.get` does not resolve.  I missed these in #5304.

File: src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java [hunk 9]
```diff
@@ -804,7 +804,7 @@ public VariantContextWriter createVCFWriter(final Path outPath) {
 
         if (createOutputVariantIndex) {
             if (null == sequenceDictionary) {
-                logger.warn("An variant index will not be created - a sequence dictionary is required to create an output index");
+                logger.warn("A variant index will not be created - a sequence dictionary is required to create an output index");
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java]: You can rest easy now!

**Comment by samuelklee** [comment]
Thanks @mwalker174, back to you!  Might need some more work depending on what you decide.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CollectAllelicCountsSpark.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CollectAllelicCountsSpark.java]: Sounds good
